workflows: use cachix consistently

wolfgangwalther opened the pull request:
This uses the `nixpkgs-ci` cachix cache in most of the CI jobs. This mostly has an effect when iterating on CI in a fork, where the target branch (and thus the merge commit) is not moving at all. It can also have a benefit when pushing to PRs targeting slower moving branches (release branches or even stable staging, which doesn't receive many commits at all). If lucky, you might also benefit from this caching if fixing a commit message or squashing commits immediately after CI has run on your PR and the target branch hasn't advanced, yet.

It also allows setting up a cachix cache in a fork, which is important to be able to test changes related to that. I am currently doing that, trying to replace some of the artifact download stuff for Eval with cachix instead (not part of this PR).

This should also make the nixpkgs-vet job a few seconds faster in general.

## Things done

- [x] Tested in my fork.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

wolfgangwalther committed:
workflows: support cachix in forks

This allows setting up a custom cachix cache in a contributor's fork.
That's most valuable when working on CI itself, testing and interating
in a fork. It's required when working on cachix related features
directly.

wolfgangwalther committed:
workflows: never push source to cachix

Pushing the source path to cachix is pointless, because the only source
we're using is Nixpkgs - and that will always be available already via
checkout. No need to ever substitute it, so no need to push it either.

wolfgangwalther committed:
workflows/lint: add cachix

For PRs targeting a fast moving branch like master, it might not
actually help because the merge commit will still be different on each
run. But on slower moving branches or when testing in a fork, it can
speed things up.

Since each of these build results is only a few hundred bytes without
any dependencies, it has essentially no cost.

Not adding this to `treefmt`, because this looks at *all* the files in
the repo, so needs to re-run even when iterating on CI only.

wolfgangwalther committed:
ci/nixpkgs-vet: memoize filesets

Speeds up evaluation a bit, because `filtered head` doesn't need to be
evaluated 3x.

wolfgangwalther committed:
ci/nixpkgs-vet: ignore .github/ and ci/ folders

nixpkgs-vet doesn't care about our CI infrastructure, so ignoring these
files will lead to more cache hits when iterating on CI related PRs.

wolfgangwalther committed:
workflows/eval: add cachix

At this stage, this only allows faster iteration when working on CI in
forks.

MattSturgeon commented on .github/workflows/build.yml:
```diff
@@ -61,9 +61,10 @@ jobs:
-          name: nixpkgs-ci
-          authToken: "${{ secrets.CACHIX_AUTH_TOKEN }}"
+          # The nixpkgs-ci cache should not be trusted or used outside of Nixpkgs CI.
+          name: ${{ vars.CACHIX_NAME || 'nixpkgs-ci' }}
```
```suggestion
          name: ${{ vars.CI_CACHIX_NAME || 'nixpkgs-ci' }}
```

?

EDIT: we already have `secrets.CACHIX_AUTH_TOKEN`, so I guess it's better to be consistent and use `vars.CACHIX_NAME`.

---

Also: is it worth adding the variable to the main nixpkgs repo, so that we don't need `||` here?

wolfgangwalther committed:
workflows/build: prevent pushing tarball to cachix

By definition the tarball changes on every commit, so it makes no sense
to cache it - it will just waste disk space.

MattSturgeon commented on .github/workflows/lint.yml:
```diff
@@ -9,6 +9,9 @@ on:
       targetSha:
         required: true
         type: string
+    secrets:
```
> Not adding this to `treefmt`, because this looks at *all* the files in the repo, so needs to re-run even when iterating on CI only.

Should we be explicit about this with a comment in the `treefmt` job/step?

MattSturgeon replied:
> I'd not add a comment

Can I ask why? If we're explicitly _not_ doing something for a reason, then to my mind this is worth highlighting to avoid forgetting this in a future PR.

> I was wondering whether there would be a way [to] cache treefmt's local DB between runs

I agree, this sounds sensible. Although maybe we'd need to index such caches to the `base` commit, to avoid trying to use out of sync DBs across various PRs?

philiptaron replied:
At work, we just invent a key that we bump whenever we'd like to, that invalidates the whole cache. It works pretty well because it's not using some proxy property like Git's hash of the source — it really means just "the root of the cache key for this purpose".

philiptaron replied:
Yeah, at work, we have a pruning job that runs every so often and therefore obviates the concern you're talking about - but in this context it makes a lot of sense.

MattSturgeon commented on ci/nixpkgs-vet.nix:
```diff
@@ -16,6 +16,9 @@ let
       root = path;
     };
+
+  base' = filtered base;
+  head' = filtered head;
```
Nice!

Is it worth naming these more explicitly, e.g:

```suggestion
  filteredBase = filtered base;
  filteredHead = filtered head;
```

```suggestion
  baseFiltered = filtered base;
  headFiltered = filtered head;
```

?

MattSturgeon reviewed (approved):
Seems good, can't see any issues.

philiptaron reviewed (commented):
Couple questions about the parameters. This looks great.

philiptaron commented on .github/workflows/codeowners-v2.yml:
```diff
@@ -64,9 +64,11 @@ jobs:
+          name: ${{ vars.CACHIX_NAME || 'nixpkgs-ci' }}
+          extraPullNames: nixpkgs-ci
+          authToken: ${{ secrets.CACHIX_AUTH_TOKEN }}
+          pushFilter: -source$
```
I'm definitely not looking for a solution or change here in this PR, but it would be amazing to fork this action for our purposes, so that we could invent a declarative syntax here instead of concatenating and copy-and-pasting regular expressions.

I also have a mild preference for the scheme used in the so-called "Magic Nix Cache" where the GitHub actions cache is used.

MattSturgeon reviewed (approved):
No issues from me
